[Merged by Bors] - feat(algebra/lie/basic): nilpotent and solvable Lie algebras

ocfnash opened the pull request:
---
An alternative proposal to my original https://github.com/leanprover-community/mathlib/pull/5241 (which I will close if we opt to merge these changes)

Commit:
Add definitions for nilpotent, solvable Lie algebras

ocfnash commented on src/algebra/lie/basic.lean:
```diff
@@ -564,8 +564,7 @@ instance lie_subalgebra_lie_ring (L' : lie_subalgebra R L) : lie_ring L' :=
 /-- A Lie subalgebra forms a new Lie algebra. -/
-instance lie_subalgebra_lie_algebra (L' : lie_subalgebra R L) :
-    @lie_algebra R L' _ (lie_subalgebra_lie_ring _ _ _) :=
+instance lie_subalgebra_lie_algebra (L' : lie_subalgebra R L) : lie_algebra R L' :=
```
This is a totally unrelated change that I just noticed.

ocfnash commented on src/algebra/lie/basic.lean:
```diff
@@ -1103,12 +1102,23 @@ variables (R : Type u) (L : Type v) (M : Type w)
+end lie_algebra
 
-/-- The derived Lie submodule of a Lie module. -/
-def derived_lie_submodule : lie_submodule R L M := ⁅(⊤ : lie_ideal R L), ⊤⁆
```
Evidently I am proposing to drop this definition (even though I only just introduced it!). This is because:

1. The name is not really right: there is no derived series for a Lie module. If anything this should be the "lower central submodule".
2. We don't need it since we should have the `derived_series` and `lower_central_series` with these changes.

jcommelin reviewed (commented):
This is looking good. But would you mind adding some stuff around these definitions?
For example:

* in an abelian lie algebra, the bracket of any two ideals is `0`.
* hence abelian lie algebras are nilpotent.
* N is nilpotent <=> \exists sequence of submodules N_i, such that [L, N_i] sub N_{i+1}, and N_0 = N, and N_i = 0 for large enough i.
* L is solvable <=> \exists sequence of ideal L_i, such that [L_i, L_i] sub L_{i+1}, and L_0 = L, and L_i = 0 for large enough i.
* hence L nilpotent implies L solvable

My hope/guess is that this might actually be not too unpleasant to formalise.

Commit:
Prove Lie bracket operation trivial for trivial Lie modules

Note that this means we thus have:
```lean
example [is_lie_abelian L] : ⁅I, J⁆ = ⊥ := by simp
```

Commit:
Trivial Lie modules are nilpotent

jcommelin commented on src/algebra/lie/basic.lean:
```diff
@@ -1222,6 +1222,9 @@ class lie_module.is_irreducible : Prop :=
 class lie_module.is_nilpotent : Prop :=
 (nilpotent : ∃ k, lie_module.lower_central_series R L M k = ⊥)
 
+lemma trivial_is_nilpotent [lie_module.is_trivial L M] : lie_module.is_nilpotent R L M :=
```
this can be an instance, right?

ocfnash replied:
Yes, thanks!

I'll push that up with the next commit, which may or may not end up arriving tomorrow :-)

Commit:
derived le lower central

Commit:
Add proof that nilpotent implies solvable

jcommelin commented on src/algebra/lie/basic.lean:
```diff
@@ -1245,6 +1245,14 @@ class lie_algebra.is_simple extends lie_module.is_irreducible R L L : Prop :=
 class lie_algebra.is_solvable : Prop :=
 (solvable : ∃ k, lie_algebra.derived_series R L k = ⊥)
 
+instance nilpotent_is_solvable [lie_module.is_nilpotent R L L] : lie_algebra.is_solvable R L :=
```
```suggestion
instance is_solvable_of_is_nilpotent [hL : lie_module.is_nilpotent R L L] : lie_algebra.is_solvable R L :=
```

jcommelin commented on src/algebra/lie/basic.lean:
```diff
@@ -1245,6 +1245,14 @@ class lie_algebra.is_simple extends lie_module.is_irreducible R L L : Prop :=
+instance nilpotent_is_solvable [lie_module.is_nilpotent R L L] : lie_algebra.is_solvable R L :=
+begin
+  rcases (lie_module.is_nilpotent.nilpotent :
+    ∃ (k : ℕ), lie_module.lower_central_series R L L k = ⊥) with ⟨k, h⟩,
```
```suggestion
  obtain ⟨k, h⟩ : ∃ k, lie_module.lower_central_series R L L k = ⊥ := hL,
```

jcommelin commented on src/algebra/lie/basic.lean:
```diff
@@ -1245,6 +1245,14 @@ class lie_algebra.is_simple extends lie_module.is_irreducible R L L : Prop :=
+begin
+  rcases (lie_module.is_nilpotent.nilpotent :
+    ∃ (k : ℕ), lie_module.lower_central_series R L L k = ⊥) with ⟨k, h⟩,
+  use k, rw ← le_bot_iff at h, rw ← le_bot_iff,
```
```suggestion
  use k, rw ← le_bot_iff at h ⊢,
```

Commit:
Update src/algebra/lie/basic.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

Commit:
Update src/algebra/lie/basic.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

Commit:
Update src/algebra/lie/basic.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

Commit:
Fix build

jcommelin reviewed (approved):
Thanks :tada:

bors merge
